src/iterators/idlist_iterator.c

```diff
@@ -9,59 +9,58 @@
 
 #include "idlist_iterator.h"
 
-static inline void setEof(IdListIterator *it, bool value) {
-  it->base.atEOF = value;
+static inline void setEof(QueryIterator *base, bool value) {
+  base->atEOF = value;
 }
 
-static inline bool isEof(const IdListIterator *it) {
-  return it->base.atEOF;
+static inline bool isEof(const QueryIterator *base) {
+  return base->atEOF;
 }
 
 static size_t IL_NumEstimated(QueryIterator *base) {
   IdListIterator *it = (IdListIterator *)base;
   return (size_t)it->size;
 }
- 
+
 /* Read the next entry from the iterator, into hit *e.
 *  Returns ITERATOR_EOF if at the end */
 static IteratorStatus IL_Read(QueryIterator *base) {
   IdListIterator *it = (IdListIterator *)base;
-  if (isEof(it) || it->offset >= it->size) {
-    setEof(it, true);
+  if (isEof(base) || it->offset >= it->size) {
+    setEof(base, true);
     return ITERATOR_EOF;
   }
 
   base->lastDocId = it->docIds[it->offset++];
-  it->base.current->docId = base->lastDocId;
+  base->current->docId = base->lastDocId;
   return ITERATOR_OK;
 }
- 
+
 /* Skip to a docid, potentially reading the entry into hit, if the docId
 * matches */
 static IteratorStatus IL_SkipTo(QueryIterator *base, t_docId docId) {
   IdListIterator *it = (IdListIterator *)base;
-  if (isEof(it) || it->offset >= it->size) {
-    setEof(it, true);
+  if (isEof(base)) {
     return ITERATOR_EOF;
   }
-
-  if (docId > it->docIds[it->size - 1]) {
-    setEof(it, true);
+  if (it->offset >= it->size || docId > it->docIds[it->size - 1]) {
+    setEof(base, true);
     return ITERATOR_EOF;
   }
 
-  int64_t top = it->size - 1, bottom = it->offset;
-  int64_t i;
+  t_offset top = MIN(it->size, it->offset + (docId - base->lastDocId) + 1);
```

Comment:
I thought about it a bit more, and we don't need the +1 here. We can add a test with a fully dense list and show that we can skip between any pair of IDs.
We should also comment about this non-trivial limit
```suggestion
  t_offset top = MIN(it->size, it->offset + (docId - base->lastDocId));
```